File: ios/AppRootRouting.swift
import Foundation

/// Root identity for the signed-in installation. A session is never enough
/// to mean "therapist" — anonymous Auth users are Patient Mode.
enum AppRootDestination: Equatable {
    case invitation
    case therapist
    case anonymousPatient
    case unauthenticated
}

enum AnonymousPatientDestination: Equatable {
    case loading
    case patientMode
    case incomplete
    case retry
}

enum AppRootRouting {
    static func destination(
        invitationActive: Bool,
        hasSession: Bool,
        isAnonymous: Bool
    ) -> AppRootDestination {
        if invitationActive { return .invitation }
        guard hasSession else { return .unauthenticated }
        if isAnonymous { return .anonymousPatient }
        return .therapist
    }

    static func anonymousDestination(
        context: AppContext?,
        isLoading: Bool
    ) -> AnonymousPatientDestination {
        if let context {
            if context.isActivePatient { return .patientMode }
            if context.role == .patient { return .incomplete }
            return .retry
        }
        if isLoading { return .loading }
        return .retry
    }
}

File: android/app/src/main/java/com/cbtipul/app/data/AppRootRouting.kt
package com.cbtipul.app.data

enum class AppRootDestination {
    Invitation,
    Therapist,
    AnonymousPatient,
    Unauthenticated,
}

enum class AnonymousPatientDestination {
    Loading,
    PatientMode,
    Incomplete,
    Retry,
}

object AppRootRouting {
    fun destination(
        invitationActive: Boolean,
        hasSession: Boolean,
        isAnonymous: Boolean,
    ): AppRootDestination {
        if (invitationActive) return AppRootDestination.Invitation
        if (!hasSession) return AppRootDestination.Unauthenticated
        return if (isAnonymous) AppRootDestination.AnonymousPatient else AppRootDestination.Therapist
    }

    fun anonymousDestination(
        context: AppContext?,
        isLoading: Boolean,
    ): AnonymousPatientDestination {
        if (context != null) {
            if (context.isActivePatient) return AnonymousPatientDestination.PatientMode
            if (context.role == AppRole.Patient) return AnonymousPatientDestination.Incomplete
            return AnonymousPatientDestination.Retry
        }
        return if (isLoading) AnonymousPatientDestination.Loading else AnonymousPatientDestination.Retry
    }
}
